### PR TITLE
array based box masks

### PR DESCRIPTION
**Pull request by ledvinap**
MSP and blackbox still need support for larger masks

### CODE REVIEW

**Comment by ledvinap** [comment]
from #2855: in MSP: 
What about using topmost bit in flightModeFlags as continuation indicator?
Topmost bit was not set in current implementation.
Store first 32 bits in current position(including 'extension' bit), then store all remaining bits at and of current packet, in 31bit chunks (+extension).
If other fields are added, they can be send after second flightModeFlags place (which may have zero length)

**Comment by ledvinap** [comment]
(array sizes are wrong, working on it)

**Comment by ledvinap** [comment]
'continuation' method for MSP implemented for review.

I am not sure if it is best possible way, would be great to get some ideas ... 

Can someone do configurator part? I don't like JS that much ...

**Comment by ledvinap** [comment]
Blackbox still logs only first 32 bits ...

**Comment by digitalentity** [comment]
This is something that's pretty much wanted in INAV as well - we are running out of box mode bits there fast.

**Comment by ledvinap** [comment]
@digitalentity : Well, int8_t is next limit ;-)

boxNames reply will eventually get too long, maybe it's time to split it now (sending 31 names per reply, adding 'page' index as parameter to MSP_BOXNAMES, defaulting to 0
And the same for permanentIds .. 

It will be backward compatible, old MSP implementation will simply ignore bits over 31 ...

**Comment by digitalentity** [comment]
Yes, this is a good idea - but not 31 names. Let's make it 32 names per reply - legacy MSP will expect uint32_t for boxIds, so equal number of names should be allowed to be reported.

However, I'd rather redo that part altogether. I'd report all box names (and IDs) supported by firmware. No more bit shuffling to figure out what modes are active. Each box mode will always have the same bitmask.

**Comment by ledvinap** [comment]
@digitalentity : as implemented now in this PR, 32th bit is used as continuation mark. We can send dummy BOX for this bit, then names/permanent IDs will match flightModeFlags exactly.
(updated) Configurator will then ignore BOX with empty name/special permanent ID

**Comment by ledvinap** [comment]
Also we can specify now that permanentBoxId is varint encoded (as in blackbox/protobuf), so that we don't ever run out of permanent IDs

**Comment by ledvinap** [comment]
@digitalentity : Current protocol allows using lots of permanentIds without making bitmask too large. Old BOXes are easy to deprecate /remove ...

**Comment by ledvinap** [comment]
Dummy entry serialization added. MSP master may use active bitmask directly, every 32nd entry will be dummy.

**Comment by ledvinap** [comment]
Well, extension bit is overkill.
I'll try to use one byte to mark number of boxid words. 
Is may be used for activeBoxids version too (allow active boxid change without reboot)

File: src/main/fc/fc_msp.c [hunk 1]
```diff
@@ -891,7 +929,12 @@ static bool mspFcProcessOutCommand(uint8_t cmdMSP, sbuf_t *dst, mspPostProcessFn
         sbufWriteU16(dst, 0);
 #endif
         sbufWriteU16(dst, sensors(SENSOR_ACC) | sensors(SENSOR_BARO) << 1 | sensors(SENSOR_MAG) << 2 | sensors(SENSOR_GPS) << 3 | sensors(SENSOR_SONAR) << 4);
-        sbufWriteU32(dst, packFlightModeFlags());
+        {
+            // pack only first 32 flags, use STATUS_EX to get them all
+            BITARRAY_TYPE(flightModeFlags, uint32_t, CHECKBOX_ITEM_COUNT);
```
Comment (DieHertz) [inline src/main/fc/fc_msp.c]: I believe configurator is currently using `MSP_STATUS` for modes tab, hence it will require more changes to it.

**Comment by hydra** [comment]
what's next on this PR?  this is currently blocking the runcam split support #2852

**Comment by DieHertz** [comment]
As no developer possesses the Split so far, it may remain blocked for a while, I'm getting a sample from RunCam in the coming weeks and will assist them with code support.
IMO this PR looks a over-engineered as of now, besides I don't see a point in implementing type-system in C for this task when we could just resort to C++.

**Comment by ledvinap** [comment]
@DieHertz : It depends .. extending to uint64_t is easier, but will only work for some time. This implementation should work for 128 bits, with easier extension in future.

Ad type system - it's just macro to define array large enough to hold required bits. Easier than typing rounding up in so many places, but can be replaced with macro that will return size only:
```
uint32_t activeBoxIds[BITARRAY_SIZE(uint32_t, CHECKBOX_ITEM_COUNT)];
```
c++ will be nice in some regards, but I'm afraid it will get quickly overused and code will be much harder to manage ...

**Comment by martinbudden** [comment]
@ledvinap , I'm inclined to agree with @DieHertz , in that this PR is somewhat over-engineered. Yes, `uint64_t` will run out at some time, but we could move to bitarray based solution if and when that happens. In the meantime a `uint64_t` based solution is simpler and easier to maintain.

As for using C++, I agree with @ledvinap - it will at some point start to cause problems and I am against a C/C++ hybrid solution. (I *would* support a total rewrite in C++, but we don't have the time and developers to do that.)

@hydra , I don't think we should rush the Runcam split support: it's important to get this right. I don't want to repeat the COLIBRI_RACE/BST situation where we have some non-mainstream code that causes problems pretty well every release.

**Comment by mikeller** [comment]
@martinbudden: Totally agreed re not adding more one-off proprietary solutions. See #3155 for an example of the trouble this gets you into.

**Comment by hydra** [comment]
i agree with the over-engineered nature of the runcam split PR.  However that is off-topic on this PR.

Right now the focus is to support > 32 bits for the 'box' flags to allow us to use the mode activation conditions for upto 64 positions.  once we have this we can continue fixing up the runcam split support.

I highly doubt that we will get past 64 any time soon.  So I'm happy to go either way, but since this blocks the other PR we need an agreement regarding the code in this PR or another solution.

**Comment by martinbudden** [comment]
@hydra , I was not stating that the RunCam Split PR was over-engineered. I was stating that the bit array used in this PR was over-engineered. That was also my interpretation of @DieHertz 's comment.

**Comment by ledvinap** [comment]
@DieHertz , @martinbudden : I agree wit you. Sometimes I don't know where to stop, feedback is very important. I'll try to simplify it (a bit) tomorrow:
- use functions to access bits (passing `void*` + index)
- `typedef` type large enough to hold all boxId_t
I hope it won't be that different from using `uint64_t` ot `typedef` directly

@hydra: runcam can be tested by temporarily disabling some boxId_t

IMO it is good idea to have lots of cheap BOXes - it is quite generic way to communicate state to configurator or trigger something. For example logical switches (openTx) can be implemented. Or OSD control and stick functions implemented using BOX (allow stick gesture to trigger box / trigger function by box. Or trigger boxes by telemetry. For example ACC calibration triggered by switch/telemetry can be quite useful)

**Comment by ledvinap** [comment]
What about MSP_STATUS_EX? (https://github.com/cleanflight/cleanflight/pull/2856/files#diff-ffa49532e0a35299ef3e307058236e8eR916). Is that OK?

I'm not sure about MSP_STATUS. It is trivial to implement it the same way as MSP_STATUS_EX (probably merging them and special-casing few fields that differ), both contain almost same data. I assume there was some historical reason for split, but I don't remember details ...

**Comment by ledvinap** [comment]
Noticed this commit: https://github.com/cleanflight/cleanflight/commit/6343da6f091dde3e5eec2cad1ac054840166a489 .
It does reuse field, but does not rename it. Is it OK to rename it to `rcModeActivationMask`?

**Comment by ledvinap** [comment]
Simplified bitarray handling and types. Should be almost equivalent to `uint64_t` variant (only typedef has to be changed and strict aliasing warning solved).

File: src/main/fc/fc_msp.c [hunk 2]
```diff
@@ -276,33 +278,53 @@ const box_t *findBoxByPermanentId(uint8_t permanentId)
...
 
-static void serializeBoxIdsReply(sbuf_t *dst)
+void serializeBoxPermanentIdFn(sbuf_t *dst, const box_t *box)
 {
+    sbufWriteU8(dst, box->permanentId);
+}
+
+// serialize 'page' of boxNames.
+// Each page contains at most 32 boxes
+static void serializeBoxReply(sbuf_t *dst, int page, serializeBoxFn *cb)
```
Comment (martinbudden) [inline src/main/fc/fc_msp.c]: I'd prefer `static void serializeBoxReply(sbuf_t *dst, int page, serializeBoxFn *serializeBox)`

File: src/main/fc/fc_msp.c [hunk 3]
```diff
@@ -869,32 +896,36 @@ static bool mspFcProcessOutCommand(uint8_t cmdMSP, sbuf_t *dst, mspPostProcessFn
...
+
         sbufWriteU16(dst, getTaskDeltaTime(TASK_GYROPID));
 #ifdef USE_I2C
         sbufWriteU16(dst, i2cGetErrorCounter());
 #else
         sbufWriteU16(dst, 0);
 #endif
         sbufWriteU16(dst, sensors(SENSOR_ACC) | sensors(SENSOR_BARO) << 1 | sensors(SENSOR_MAG) << 2 | sensors(SENSOR_GPS) << 3 | sensors(SENSOR_SONAR) << 4);
-        sbufWriteU32(dst, packFlightModeFlags());
+            sbufWriteU32(dst, ((uint32_t*)&flightModeFlags)[0]);      // unconditional part of flags
```
Comment (martinbudden) [inline src/main/fc/fc_msp.c]: Whitespace.

File: src/main/msp/msp.h [hunk 4]
```diff
@@ -23,7 +23,8 @@
 typedef enum {
     MSP_RESULT_ACK = 1,
     MSP_RESULT_ERROR = -1,
-    MSP_RESULT_NO_REPLY = 0
+    MSP_RESULT_NO_REPLY = 0,
+    MSP_RESULT_CMDUNKNOWN = -2,   // don't know how to process command, try next handler
```
Comment (martinbudden) [inline src/main/msp/msp.h]: I prefer `MSP_RESULT_CMD_UNKNOWN`

File: src/main/fc/fc_msp.c [hunk 5]
```diff
@@ -869,32 +896,36 @@ static bool mspFcProcessOutCommand(uint8_t cmdMSP, sbuf_t *dst, mspPostProcessFn
...
-        sbufWriteU16(dst, 0);
-#endif
-        sbufWriteU16(dst, sensors(SENSOR_ACC) | sensors(SENSOR_BARO) << 1 | sensors(SENSOR_MAG) << 2 | sensors(SENSOR_GPS) << 3 | sensors(SENSOR_SONAR) << 4);
-        sbufWriteU32(dst, packFlightModeFlags());
-        sbufWriteU8(dst, getCurrentPidProfileIndex());
-        sbufWriteU16(dst, constrain(averageSystemLoadPercent, 0, 100));
-        sbufWriteU16(dst, 0); // gyro cycle time
+            // write flightModeFlags header. Lowest 4 bits contain number of bytes that follow
+            // header is emited even when all bits fit into 32 bits to allow future extension
+            int cnt = (flagBits - 32 + 7) / 8;  // 32 already stored, round up
```
Comment (martinbudden) [inline src/main/fc/fc_msp.c]: I prefer `count`

File: src/main/fc/fc_msp.c [hunk 6]
```diff
@@ -110,12 +112,13 @@
 #endif
 
 #define STATIC_ASSERT(condition, name) \
-    typedef char assert_failed_ ## name [(condition) ? 1 : -1 ]
+    typedef char assert_failed_ ## name [(condition) ? 1 : -1 ] __attribute__((unused))
```
Comment (martinbudden) [inline src/main/fc/fc_msp.c]: `STATIC_ASSERT` could be move into `utils.h`

File: src/main/fc/fc_msp.c [hunk 7]
```diff
@@ -869,32 +895,36 @@ static bool mspFcProcessOutCommand(uint8_t cmdMSP, sbuf_t *dst, mspPostProcessFn
...
-#endif
-        sbufWriteU16(dst, sensors(SENSOR_ACC) | sensors(SENSOR_BARO) << 1 | sensors(SENSOR_MAG) << 2 | sensors(SENSOR_GPS) << 3 | sensors(SENSOR_SONAR) << 4);
-        sbufWriteU32(dst, packFlightModeFlags());
-        sbufWriteU8(dst, getCurrentPidProfileIndex());
-        sbufWriteU16(dst, constrain(averageSystemLoadPercent, 0, 100));
-        sbufWriteU16(dst, 0); // gyro cycle time
+            sbufWriteU16(dst, 0);
+#endif
+            sbufWriteU16(dst, sensors(SENSOR_ACC) | sensors(SENSOR_BARO) << 1 | sensors(SENSOR_MAG) << 2 | sensors(SENSOR_GPS) << 3 | sensors(SENSOR_SONAR) << 4);
+            sbufWriteU32(dst, ((uint32_t*)&flightModeFlags)[0]);      // unconditional part of flags
```
Comment (DieHertz) [inline src/main/fc/fc_msp.c]: Why not just `&flightModeFlags.bits[0]` or `flightModeFlags.bits`?

Comment (ledvinap) [inline src/main/fc/fc_msp.c]: I wanted the type to be as opaque as possible. Using struct content will be implementation dependent.
One option is to use `union` to mask implementation details.

`sbufWriteData` is probably better

File: src/main/fc/fc_msp.c (hunk 6 above)
Comment (ledvinap) [inline src/main/fc/fc_msp.c]: I know, I'll fix it in separate PR (removing BUILD_BUG_ON,  STATIC_ASSERT can report error message)

File: src/main/fc/fc_msp.c (hunk 7 above)
Comment (DieHertz) [inline src/main/fc/fc_msp.c]: Current variant assumes `boxBitmask_t` can be interpreted as an array of at-least one `uint32_t`, hence not really opaque. I'd use a union at least.

Comment (ledvinap) [inline src/main/fc/fc_msp.c]: now:
```
sbufWriteData(dst, &flightModeFlags, 4);        // unconditional part of flags, first 32 bits
```
seems reasonable to me - flightModeFlags is unstructured bit bit array, larger than 32 bits (I can add check for size, but I doubt it's usefull)

Comment (DieHertz) [inline src/main/fc/fc_msp.c]: Looks good

**Comment by martinbudden** [comment]
@ledvinap , can you ensure test code runs?

**Comment by ledvinap** [comment]
@martinbudden : Of course, just need to find some time ...

**Comment by ledvinap** [comment]
rc_controls split to simplify unittest dependency hell a bit

**Comment by mikeller** [comment]
Erm, either I've gone blind, your you've forgotten to check in `rc_modes.[ch]`.

**Comment by ledvinap** [comment]
@mikeller: sorry..

**Comment by ledvinap** [comment]
Files added

**Comment by ledvinap** [comment]
PR build without errors now ...

**Comment by mikeller** [comment]
If you think it's ready, would you mind a rebase?

**Comment by ledvinap** [comment]
I'll squash commits to get clean history. 
I hope I didn't miss something..

**Comment by hydra** [comment]
great, i should have some time to review and merge this weekend if it's ready.

**Comment by mikeller** [comment]
Would be appreciated, there is one candidate for an additional box in Betaflight already.

**Comment by ledvinap** [comment]
Squashed, ready to merge

TODO:
- add configurator support (it will work now when fewer than 32 bits are active)
- blackbox logs only first 32 bits of activeRcmodes (and names them wrong and wastes blackbox memory)

**Comment by hydra** [comment]
@mikeller for the candidate in a BF PR that needs a 'box' please re-use BOXGOV as it's currently unused by the codebase.

**Comment by hydra** [comment]
good work everyone merging.

@ledvinap can you do another PR for the second TODO you mentioned above?  If you don't have time please create a new issue so we don't forget this.

**Comment by mikeller** [comment]
@hydra: When re-using, do I just assign a new permanent ID? I want to avoid screwing with modes in backup / restore, like what happened the last time we reused boxes.

**Comment by ledvinap** [comment]
@mikeller : As designed, permament ID + name is only thing visible over MSP. Actual bit position is dependent on configuration and features and enum value is never sent over MSP.
Configs (cli dump) should store permanentId, but I did check this only briefly.

**Pull request merged**